### PR TITLE
Avoid unnecessary copies of  endpoints object in kube-proxy

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

File: pkg/proxy/config/config.go [hunk 1]
```diff
@@ -63,7 +63,13 @@ type EndpointsConfigHandler interface {
 	// OnEndpointsUpdate gets called when endpoints configuration is changed for a given
 	// service on any of the configuration sources. An example is when a new
 	// service comes up, or when containers come up or down for an existing service.
-	OnEndpointsUpdate(endpoints []api.Endpoints)
+	//
```
Comment (bowei) [inline pkg/proxy/config/config.go]: You should probably state explicitly that it is being passed by reference for efficiency, it won't be obvious to other people reading the code

**Comment by bowei** [comment]
what is the perf impact of this change?

**Comment by wojtek-t** [comment]
> what is the perf impact of this change?

I'm focusing first on the "api-related" part of kube-proxy (for now only this part is reuse in kubemark) - by api-related thing I mean what we do before entering proxier (i.e. watching from apiserver etc.)
For this part https://github.com/kubernetes/kubernetes/pull/42108 reduce this part by ~2/3rd, and this PR (together with the sibling one, I'm going to do for services, though didn't yet fully verified that it is safe, so didn't yet do that), will reduce it by the remaining 1/3rd. So those 3 PRs together will reduce the cost of this part from lineral to constant (per endpoint or service update). Once this is done, I/we will look into proxier part, but generally those are kind of independent optimizations.
Looking into whole kubeeproxy, this isn't probably very significant percentage, but if we go through all the code in kube-proxy, I'm sure we can optimize a lot.

**Comment by wojtek-t** [comment]
Once we're done with those low-hanging fruits, I will look deeper into proxier and then we can probably meet and discuss potential improvements.

**Comment by bowei** [comment]
ok sounds good (with the comment change)

File: pkg/proxy/config/config.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/proxy/config/config.go]: done

**Comment by wojtek-t** [comment]
@k8s-bot kops aws test this, issue: #IGNORE (cluster didn't started)

**Comment by wojtek-t** [comment]
Trivial rebase - reaaplying label.

**Comment by wojtek-t** [comment]
Just rebased.

**Pull request merged**